File: Java/LinkedList/03_Singly_Circular_LinkedList_Java/SinglyCL_002.java
/*
===============================================================================
Program Name    : Creation and Linking of Three Nodes in Singly Circular Linked List
Program Number  : 002
Description     : This program demonstrates the creation of three nodes using
                  the default constructor in Java. The data members are
                  initialized separately, and the nodes are linked together
                  using reference variables to form a simple linked list.

Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class and Object
2. Default Constructor
3. Self-Referential Class
4. Dynamic Memory Allocation using new
5. Reference Variables
6. Node Linking

Key Learning Points:
- Objects can be created using a default constructor.
- Data members can be assigned values after object creation.
- Each node stores both data and the address of the next node.
- Nodes are connected through reference variables.
- The last node points to NULL, indicating the end of the list.

===============================================================================
*/
// Program 02 : Creation of Three Nodes

class node
{
    public int data;
    public node next;

    public node()
    {
        data = 0;
        next = null;
    }
}

class SinglyCL2
{
    public static void main(String A[])
    {
        node first = null;
        node second = null;
        node third = null;

       
        first = new node();
        second = new node();
        third = new node();

       
        first.data = 11;
        second.data = 21;
        third.data = 51;

        
        first.next = second;
        second.next = third;
        third.next = null;

        
        System.out.println(first.data);
        System.out.println(second.data);
        System.out.println(third.data);
    }
}
/*
===============================================================================
Expected Output:
===============================================================================

11
21
51

===============================================================================
Memory Representation:
===============================================================================

        first                  second                  third
          │                       │                       │
          ▼                       ▼                       ▼
     +-----------+          +-----------+          +-----------+
     | data = 11 |          | data = 21 |          | data = 51 |
     | next -----|--------> | next -----|--------> | next=NULL |
     +-----------+          +-----------+          +-----------+

===============================================================================
Explanation:
===============================================================================

The program defines a self-referential class named node containing two data
members: data and next.

Three node objects are created dynamically using the default constructor.

Initially, each node contains the default values:
- data = 0
- next = null

After object creation, values 11, 21, and 51 are assigned to the respective
nodes.

The nodes are then connected as follows:

first → second → third → NULL

Finally, the data stored in each node is displayed using the corresponding
reference variables.

===============================================================================
Program Flow:
===============================================================================

Start Program
      │
      ▼
Create Three Reference Variables
      │
      ▼
Create Three Node Objects
      │
      ▼
Assign Data to Each Node
      │
      ▼
Link Nodes Together
      │
      ▼
Set Last Node Next as NULL
      │
      ▼
Display Node Data
      │
      ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Defined a self-referential node class.
2. Declared three node reference variables.
3. Created three node objects using the default constructor.
4. Assigned values to each node.
5. Linked all three nodes together.
6. Set the last node reference to NULL.
7. Displayed the data stored in each node.

===============================================================================
Time Complexity:
===============================================================================

Node Creation    : O(1)
Node Linking     : O(1)
Display Data     : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 002 demonstrates the creation of multiple nodes using the default
constructor and shows how values can be assigned after object creation.
The program also illustrates how nodes are connected to form a linked
structure, providing the basic foundation for implementing Singly Circular
Linked Lists in Java.

===============================================================================
*/

File: C++/LinkedList/03_Singly_Circular_Linear_LinkedList_CPP/SinglyCL_002.cpp
/*
===============================================================================
Program Name    : Singly Circular Linked List - Creation of Two Nodes
Program Number  : 002
Description     : This program demonstrates how to create two nodes dynamically
                  in C++. It allocates memory using the new operator, stores
                  data in each node, and connects them to form the initial
                  circular linked list.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Dynamic Memory Allocation
2. Self-Referential Structure
3. Pointers
4. new Operator
5. Circular Node Linking
6. Pointer Assignment

Key Learning Points:
- Nodes can be created dynamically using the new operator.
- Each node stores data and the address of the next node.
- The last node points back to the first node.
- Circular linking eliminates the NULL pointer at the end.
- This is the simplest form of a Singly Circular Linked List.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

int main()
{
    PNODE first = NULL;
    PNODE second = NULL;

    first = new NODE;
    second = new NODE;

    first->data = 11;
    second->data = 21;

    first->next = second;
    second->next = first;

    cout<<"First Node Data  : "<<first->data<<endl;
    cout<<"Second Node Data : "<<second->data<<endl;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

First Node Data  : 11
Second Node Data : 21

===============================================================================
Memory Representation:
===============================================================================

                 first
                   │
                   ▼
        +-----------+-----------+
        |   11      |     •-----|------------------+
        +-----------+-----------+                  |
                                                   |
                                                   ▼
                                          +-----------+-----------+
                                          |   21      |     •-----|
                                          +-----------+-----------+
                                                   ▲
                                                   |
                                                   +----------------+

Both nodes form a circular chain.

===============================================================================
Pointer Explanation:
===============================================================================

first
└── Stores the address of the first node.

second
└── Stores the address of the second node.

first->next
└── Points to the second node.

second->next
└── Points back to the first node, completing the circle.

===============================================================================
Program Flow:
===============================================================================

Create First Node
        │
        ▼
Create Second Node
        │
        ▼
Store Data
        │
        ▼
Connect First → Second
        │
        ▼
Connect Second → First
        │
        ▼
Display Node Data

===============================================================================
Time Complexity:
===============================================================================

Node Creation      : O(1)
Node Linking       : O(1)
Display Data       : O(1)

Space Complexity   : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 002 demonstrates the creation of two dynamically allocated nodes and
their circular connection. Unlike a Singly Linear Linked List, the last node
does not point to NULL; instead, it points back to the first node, forming a
closed circular structure. This is the first practical implementation of a
Singly Circular Linked List.

===============================================================================
*/
